### PR TITLE
First version of print file generator

### PR DESCRIPTION
**Pull request by AdamHawtin**
## Motivation and Context
We want to be able to process a sample file and generate initial print files without having to run the whole of RM. This will require modifications to the RM services and sample loader to be able to ingest the processed sample file back into RM later.

Performance wise this script takes around 70 seconds per hundred thousand sample units on my machine. This is limited by having to fetch IAC's for every sample unit.

This first version uses a placeholder uuid for `QID` and has mappings for England and Wales initial contact letters. Adding more initial contact letters with the same template should be trivial, just add the treatment country to productpack code mapping. 

The script does not yet produce a print manifest file.

## What has Changed
- Add process sample file script
- Add generate print files script
- Add unit tests

## How to Test
Run a sample file through the process sample file script, then the processed file through the generate print files script to get the print files. Check the print files match the print file spec.

## Links
https://trello.com/c/gzaskio6/647-investigate-script-to-generate-print-files-from-sample-file-directly

### CODE REVIEW

**Comment by AdamHawtin** [comment]
[Investigate script to generate print files from sample file directly](https://trello.com/c/gzaskio6/647-investigate-script-to-generate-print-files-from-sample-file-directly)

File: iac_controller.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        self._max_total_iacs = max_total_iacs
+        self._batch_size = batch_size
+        self._total_iacs_fetched = 0
+
+    def get_iac(self) -> str:
+        if not self._iac_pool:
+            self._add_iac_batch_to_pool()
+        return self._iac_pool.pop()
+
+    def _add_iac_batch_to_pool(self):
```
Comment (NickGrantX) [inline iac_controller.py:24]: could this method be async so that it wouldn't cause delays when it's topping up the pool?

Comment (AdamHawtin) [reply iac_controller.py:24]: I'll have a look at doing that but I think the time spent putting the codes on the queue is negligible compared to the time spent in the HTTP call.

Comment (LukeLoze1) [reply iac_controller.py:24]: For what it's worth I would keep it simple for now.  There's not a problem with speed at the moment.

File: process_sample_file.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        sample_file, sample_file_line_counter = tee(sample_file)
+        sample_size = sum(1 for _ in sample_file_line_counter) - 1
+        process_sample_file_rows(sample_file, sample_size, output_file_path)
+
+
+def process_sample_file_rows(sample_file: Iterable[str], sample_size, output_file_path):
+    print(f'Preparing to process {sample_size} sample units')
+
+    sample_file_reader = csv.DictReader(sample_file, delimiter=',')
+    iac_controller = IACController(max_total_iacs=sample_size)
```
Comment (LukeLoze1) [inline process_sample_file.py:23]: Nice/Thumbs up etc

File: progress.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import sys
+
+
+def print_sample_units_progress(number_processed, total):
```
Comment (LukeLoze1) [inline progress.py]: as discussed this function is nice, but could do with comments or ideally variable naming to make it clear what it's doing and why

**Review by LukeLoze1 — APPROVED** [review]

File: iac_controller.py (hunk 1 above)
Comment (AdamHawtin) [reply iac_controller.py:24]: I think I misunderstood you in my first reply, but I agree with Luke. If we find we need this script to be faster in the future that's probably the first thing to try.

**Review by liamtoozer — APPROVED** [review]
Nice!

File: iac_controller.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        count = self._batch_size if self._batch_size < self._max_total_iacs - self._total_iacs_fetched\
```
Comment (jcox-dev) [inline iac_controller.py]: A comment prompt here about how this only fetches IACs in batches smaller than the remaining total required would be helpful for those trying to grok what this (long) ternary line is doing.

File: tests/test_iac_controller.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import responses
```
Comment (jcox-dev) [inline tests/test_iac_controller.py:3]: I might suggest using something similar to the following to ensure that the count mechanism of the controller is working as expected:

```python
import json

import responses

from iac_controller import IACController
from tests import IACConfigTestCase


class TestIACController(IACConfigTestCase):
    iac_batches = {'0': [], '1': ['testiaccode3'], '2': ['testiaccode1', 'testiaccode2']}

    @staticmethod
    def request_callback(request):
        payload = json.loads(request.body)
        resp_body = TestIACController.iac_batches[payload['count']]
        return (200, {}, json.dumps(resp_body))

    def test_get_iac(self):
        # Given
        with responses.RequestsMock(assert_all_requests_are_fired=True) as rsps:
            rsps.add_callback(
                responses.POST, f'{self.test_iac_url}/iacs',
                callback=self.request_callback,
                content_type='application/json'
            )
            iac_controller = IACController(max_total_iacs=1, batch_size=1)

            # When
            iac = iac_controller.get_iac()

        # Then
        assert iac == self.iac_batches['1'][0]

    def test_get_iac_multiple_batches(self):

        # Given
        with responses.RequestsMock(assert_all_requests_are_fired=True) as rsps:
            rsps.add_callback(
                responses.POST, f'{self.test_iac_url}/iacs',
                callback=self.request_callback,
                content_type='application/json'
            )
            iac_controller = IACController(max_total_iacs=3, batch_size=2)

            # When
            first_iac = iac_controller.get_iac()

            second_iac = iac_controller.get_iac()

            third_iac = iac_controller.get_iac()

        # Then
        assert first_iac in self.iac_batches['2']
        assert second_iac in self.iac_batches['2']
        assert third_iac == self.iac_batches['1'][0]
```

File: iac_controller.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import os
+from collections import deque
+from typing import Collection
+
+import requests
+from requests.auth import HTTPBasicAuth
+
+
+class IACController:
```
Comment (jcox-dev) [inline iac_controller.py:9]: Liking the design of this one 👍 

Fill pool (with a request to the IAC service) to drain from then refill when empty. Nice.

**Review by jcox-dev — CHANGES_REQUESTED** [review]
Tests all pass and code is clean and does what it says on the tin. 

Just a comment about readability of a line and potential improved test coverage of the controller's functionality.

File: iac_controller.py (hunk 4 above)
Comment (AdamHawtin) [reply iac_controller.py]: The line is a bit long, I'm going to split it out into a function so I can name it.

**Review by jcox-dev — APPROVED** [review]

**Pull request merged**